sale_order_product_recommendation/wizards/sale_order_recommendation.py

```diff
@@ -0,0 +1,177 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Jairo Llopis <jairo.llopis@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from datetime import datetime, timedelta
+from openerp import api, fields, models
+
+
+class SaleOrderRecommendation(models.TransientModel):
+    _name = "sale.order.recommendation"
+    _description = "Recommended products for current sale order"
+
+    order_id = fields.Many2one(
+        "sale.order",
+        "Sale Order",
+        default=lambda self: self._default_order_id(),
+        required=True,
+        readonly=True,
+        ondelete="cascade",
+    )
+    months = fields.Float(
+        default=6,
+        required=True,
+        help="Consider these months backwards to generate recommendations.",
+    )
+    line_ids = fields.One2many(
+        "sale.order.recommendation.line",
+        "wizard_id",
+        "Products",
+    )
+    line_amount = fields.Integer(
+        "Number of recommendations",
+        default=15,
+        required=True,
+        help="The less, the faster they will be found.",
+    )
+
+    @api.model
+    def _default_order_id(self):
+        return self.env.context.get("active_id", False)
+
+    @api.multi
+    @api.onchange("order_id", "months", "line_amount")
+    def _generate_recommendations(self):
+        """Generate lines according to context sale order."""
+        start = datetime.now() - timedelta(days=self.months * 30)
+        start = fields.Datetime.to_string(start)
+        order_lines = self.order_id.order_line
+        existing_product_ids = set(order_lines.mapped("product_id").ids)
+        self.line_ids = False
+        # Search delivered products in previous months
+        found_lines = self.env["sale.order.line"].read_group(
+            [
+                ("order_partner_id", "child_of",
+                 self.order_id.partner_id.commercial_partner_id.id),
+                ("order_id.date_order", ">=", start),
```

Comment:
order_id.date_order can be replace with create_date (if little diferences are assumed) or create related field and set index=True to better performance

Reply:
Execute Sql sentence with OVER clause can be another option to performance